pkg/aggregator/context_resolver.go

```diff
@@ -42,30 +43,39 @@ type contextResolver struct {
 	keyGenerator  *ckey.KeyGenerator
 	taggerBuffer  *tagset.HashingTagsAccumulator
 	metricBuffer  *tagset.HashingTagsAccumulator
+	limiter       *limiter.Limiter
 }
 
 // generateContextKey generates the contextKey associated with the context of the metricSample
 func (cr *contextResolver) generateContextKey(metricSampleContext metrics.MetricSampleContext) (ckey.ContextKey, ckey.TagsKey, ckey.TagsKey) {
 	return cr.keyGenerator.GenerateWithTags2(metricSampleContext.GetName(), metricSampleContext.GetHost(), cr.taggerBuffer, cr.metricBuffer)
 }
 
-func newContextResolver(cache *tags.Store) *contextResolver {
+func newContextResolver(cache *tags.Store, limiter *limiter.Limiter) *contextResolver {
 	return &contextResolver{
 		contextsByKey: make(map[ckey.ContextKey]*Context),
 		countsByMtype: make([]uint64, metrics.NumMetricTypes),
 		tagsCache:     cache,
 		keyGenerator:  ckey.NewKeyGenerator(),
 		taggerBuffer:  tagset.NewHashingTagsAccumulator(),
 		metricBuffer:  tagset.NewHashingTagsAccumulator(),
+		limiter:       limiter,
 	}
 }
 
 // trackContext returns the contextKey associated with the context of the metricSample and tracks that context
-func (cr *contextResolver) trackContext(metricSampleContext metrics.MetricSampleContext) ckey.ContextKey {
-	metricSampleContext.GetTags(cr.taggerBuffer, cr.metricBuffer)                  // tags here are not sorted and can contain duplicates
+func (cr *contextResolver) trackContext(metricSampleContext metrics.MetricSampleContext) (ckey.ContextKey, bool) {
+	metricSampleContext.GetTags(cr.taggerBuffer, cr.metricBuffer) // tags here are not sorted and can contain duplicates
+	defer cr.taggerBuffer.Reset()
+	defer cr.metricBuffer.Reset()
+
 	contextKey, taggerKey, metricKey := cr.generateContextKey(metricSampleContext) // the generator will remove duplicates (and doesn't mind the order)
 
 	if _, ok := cr.contextsByKey[contextKey]; !ok {
+		if !cr.canAdd() {
```

Comment:
```suggestion
		if !cr.tryAdd() {
```
maybe a slightly clearer name, its not obvious (to me) that `canAdd` will also actually track the context if its under the limit